internal/pkg/certificates/certificates.go

```diff
@@ -0,0 +1,148 @@
+package certificates
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
+	controllerCrypto "github.com/kubewarden/kubewarden-controller/internal/pkg/crypto"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	logger = ctrl.Log.WithName("certificates")
+)
+
+// This function will create and store in secrets, when necessary, the root CA
+// and the certificate used in the controller. If the certificate is created,
+// the process exit to force kubernetes to redeploy ensuring that the latest
+// version of the certificates will be in use
+func CheckRootCA(ctx context.Context, mgr ctrl.Manager, deploymentsNamespace string, controllerWebhookServiceName string) error {
+	logger.V(2).Info("Checking root certificate authority")
+	k8s := mgr.GetClient()
+	rootCASecret := &corev1.Secret{}
+	err := k8s.Get(ctx, client.ObjectKey{Namespace: deploymentsNamespace, Name: constants.RootCASecretName}, rootCASecret)
+	if apierrors.IsNotFound(err) {
+		rootCA, err := createRootCa(ctx, k8s, rootCASecret, deploymentsNamespace)
+		if err != nil {
+			return errors.Join(errors.New("unable to create root CA secret"), err)
+		}
+		logger.Info("Root CA secret created")
+		err = renewControllerCerts(ctx, k8s, rootCA, []string{fmt.Sprintf("%s.%s.svc.cluster.local", controllerWebhookServiceName, deploymentsNamespace), fmt.Sprintf("%s.%s.svc", controllerWebhookServiceName, deploymentsNamespace)}, deploymentsNamespace)
+		if err != nil {
+			return err
+		}
+		logger.Info("Start to update webhook with latest CA")
+		certificatePEMEncoded, err := rootCA.PEMEncodeCertificate()
+		if err != nil {
+			return errors.Join(errors.New("unable to PEM encode CA certificate"), err)
+		}
+		err = updateWebhooksCaBundle(ctx, k8s, certificatePEMEncoded.Bytes())
+		if err != nil {
+			return err
+		}
+		logger.Info("Certificates and webhooks updated")
+		logger.Info("Restarting controller due a root certificate authority renewal")
+		os.Exit(constants.ControllerReturnCodeCAInitialized)
+		return nil
```

Comment:
We can leave that out (and update the function documentation accordingly):

* We are going to create the certificate before starting the webhook server
* The webhook server provided by controller runtime can detect changes done to the certificate & its key. We just need to update the Secret, kubelet will sync the contents on the files mounted inside of the Pod, controller-runtime will reload them and start answering with the new cert